docs/markdown/Writing Plugins/tutorials/01-create-new-goal-with-a-custom-target.md

```diff
@@ -0,0 +1,345 @@
+
+```python pants-plugins/project_version/rules.py
+from dataclasses import dataclass
+
+from pants.engine.console import Console
+from pants.engine.fs import DigestContents
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.internals.native_engine import Digest
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.rules import collect_rules, goal_rule, rule
+from pants.engine.target import (HydratedSources, HydrateSourcesRequest,
+                                 SourcesField, Targets)
+from project_version.target_types import ProjectVersionTarget
+
+
+@dataclass(frozen=True)
+class ProjectVersionFileView:
+    path: str
+    version: str
+
+
+@rule
+async def get_project_version_file_view(
+    target: ProjectVersionTarget,
+) -> ProjectVersionFileView:
+    sources = await Get(HydratedSources, HydrateSourcesRequest(target[SourcesField]))
+    digest_contents = await Get(DigestContents, Digest, sources.snapshot.digest)
+    if len(digest_contents) > 1:
+        raise ValueError("There should be only one source file.")
+    file_content = digest_contents[0]
+    return ProjectVersionFileView(
+        path=file_content.path, version=file_content.content.decode("utf-8").strip()
+    )
+
+
+class ProjectVersionSubsystem(GoalSubsystem):
+    name = "project-version"
+    help = "Show representation of the project version from the `VERSION` file."
+
+
+class ProjectVersionGoal(Goal):
+    subsystem_cls = ProjectVersionSubsystem
+
+
+@goal_rule
+async def goal_show_project_version(
+    console: Console, targets: Targets
+) -> ProjectVersionGoal:
+    targets = [tgt for tgt in targets if tgt.alias == ProjectVersionTarget.alias]
+    results = await MultiGet(
+        Get(ProjectVersionFileView, ProjectVersionTarget, target) for target in targets
+    )
+    for result in results:
+        console.print_stdout(str(result))
+    return ProjectVersionGoal(exit_code=0)
+
+
+def rules():
+    return collect_rules()
+```
+```python pants-plugins/project_version/targets.py
+from pants.engine.target import COMMON_TARGET_FIELDS, SingleSourceField, Target
+
+
+class ProjectVersionTarget(Target):
+    alias = "version_file"
+    core_fields = (*COMMON_TARGET_FIELDS, SingleSourceField)
+    help = "A project version target representing the VERSION file."
+```
+```python pants-plugins/project_version/register.py
+from typing import Iterable
+
+import project_version.rules as project_version_rules
+from pants.engine.target import Target
+from project_version.target_types import ProjectVersionTarget
+
+
+def target_types() -> Iterable[type[Target]]:
+    return [ProjectVersionTarget]
+
+
+def rules():
+    return [*project_version_rules.rules()]
+```
+
+Running our goal:
+
+```
+$ /pants project-version myapp
+ProjectVersionFileView(path='myapp/VERSION', version='0.0.1')
+```
+
+The `VERSION` file was read and its contents is shown in the console. You have now finished writing your plugin!
```

Comment:
Maybe conclude with a few words about possible future direction for this plugin.. ? :)

Reply:
Yes, fixed! :)